src/hotspot/share/classfile/classLoaderData.cpp

```diff
@@ -300,6 +300,30 @@ bool ClassLoaderData::try_claim(int claim) {
   }
 }
 
+void ClassLoaderData::demote_strong_roots() {
+  // The oop handle area contains strong roots that the GC traces from. We are about
+  // to demote them to strong roots that the GC does *not* trace from. Conceptually,
+  // we are retiring a normal strong root, and creating a strong non-root handle, which
+  // happens to reuse the same address as the normal strong root had.
+  // The way we would retire a strong root, is by clearing it. Then its address can be
+  // reused for a new handle, that isn't a normal strong root. We do that dance below.
+  class TransitionRootsOopClosure : public OopClosure {
+  public:
+    virtual void do_oop(oop* p) {
+      oop obj = NativeAccess<>::oop_load(p); // Load the strong root
+      NativeAccess<>::oop_store(p, nullptr); // Clear the strong root
+      // No-op free handle
+      // No-op allocate new handle using the same address
+      NativeAccess<>::oop_store(p, obj); // Store the strong non-root
```

Comment:
Is there an issue if someone with an `OopHandle(p)` racingly reads null while a root is transitioning?

Reply:
Shouldn't be, but I'll remove the null store and write some more comments, to be on the safe side.